// tdesign-vue-next/es/upload/image.mjs
/**
 * tdesign v0.21.1
 * (c) 2022 tdesign
 * @license MIT
 */

import { _ as _defineProperty } from '../_chunks/dep-0b4c3c44.mjs';
import { defineComponent, computed, createVNode } from 'vue';
import { AddIcon, DeleteIcon, BrowseIcon } from 'tdesign-icons-vue-next';
import { Loading } from '../loading/index.mjs';
import props from './props.mjs';
import { useFormDisabled } from '../form/hooks.mjs';
import { useGlobalIcon } from '../hooks/useGlobalIcon.mjs';
import { usePrefixClass, useCommonClassName } from '../hooks/useConfig.mjs';
import { useConfig } from '../config-provider/useConfig.mjs';
import '../loading/directive.mjs';
import '../loading/plugin.mjs';
import '../loading/loading.mjs';
import '../_chunks/dep-dcf624da.mjs';
import '../_chunks/dep-556a2f2c.mjs';
import '../loading/icon/gradient.mjs';
import '../_common/js/loading/circle-adapter.mjs';
import '../_common/js/utils/set-style.mjs';
import '../_common/js/utils/helper.mjs';
import '../_chunks/dep-8c39d78a.mjs';
import '../_chunks/dep-e68fd68c.mjs';
import '../_chunks/dep-bfb23a5e.mjs';
import '../_chunks/dep-a48de419.mjs';
import '../_chunks/dep-2a7dc5c7.mjs';
import '../_chunks/dep-9311bf66.mjs';
import '../_chunks/dep-4889eb51.mjs';
import '../_chunks/dep-0b4185fe.mjs';
import '../_chunks/dep-3ca5fb38.mjs';
import '../_chunks/dep-7071345f.mjs';
import '../_chunks/dep-d2403b2f.mjs';
import '../_chunks/dep-7922050a.mjs';
import '../_chunks/dep-169ea640.mjs';
import '../_chunks/dep-f6b25712.mjs';
import '../_chunks/dep-59545a4d.mjs';
import '../_chunks/dep-c27b1c14.mjs';
import '../_chunks/dep-7d96367f.mjs';
import '../_chunks/dep-827c4600.mjs';
import '../_chunks/dep-0820e0e2.mjs';
import '../_chunks/dep-caad2e91.mjs';
import '../_chunks/dep-3d4aad09.mjs';
import '../_chunks/dep-8e38f523.mjs';
import '../_chunks/dep-469bf54b.mjs';
import '../_chunks/dep-8ef9fb8e.mjs';
import '../_chunks/dep-e769aafd.mjs';
import '../_chunks/dep-400798c5.mjs';
import '../_chunks/dep-361522ee.mjs';
import '../_common/js/global-config/default-config.mjs';
import '../_common/js/global-config/locale/zh_CN.mjs';
import '../config-provider/type.mjs';
import '../utils/dom.mjs';
import '../_chunks/dep-2684fd15.mjs';
import '../utils/easing.mjs';
import '../utils/render-tnode.mjs';
import '../_chunks/dep-40522c7c.mjs';
import '../_chunks/dep-83862ee1.mjs';
import '../_chunks/dep-8e451077.mjs';
import '../_chunks/dep-a82b01df.mjs';
import '../_chunks/dep-192e10f1.mjs';
import '../_chunks/dep-977c5cdd.mjs';
import '../_chunks/dep-755716aa.mjs';
import '../_chunks/dep-82798823.mjs';
import '../_chunks/dep-52cbb58a.mjs';
import '../_chunks/dep-db11154e.mjs';
import '../_chunks/dep-baa4f9dc.mjs';
import '../utils/transfer-dom.mjs';
import '../loading/props.mjs';
import '../utils/withInstall.mjs';
import './style/css.mjs';
import '../loading/type.mjs';

var ImageCard = defineComponent({
  name: "TImageUpload",
  props: {
    files: props.files,
    loadingFile: {
      type: Object,
      "default": function _default() {
        return null;
      }
    },
    locale: {
      type: Object,
      "default": function _default() {
        return null;
      }
    },
    percent: Number,
    showUploadProgress: props.showUploadProgress,
    placeholder: props.placeholder,
    multiple: props.multiple,
    max: props.max,
    disabled: props.disabled,
    onClick: Function,
    onRemove: Function,
    onImgPreview: Function
  },
  setup: function setup(props2) {
    var disabled = useFormDisabled();

    var _useConfig = useConfig("upload"),
        prefix = _useConfig.classPrefix,
        globalConfig = _useConfig.globalConfig;

    var UPLOAD_NAME = usePrefixClass("upload");

    var _useGlobalIcon = useGlobalIcon({
      AddIcon: AddIcon,
      DeleteIcon: DeleteIcon,
      BrowseIcon: BrowseIcon
    }),
        AddIcon$1 = _useGlobalIcon.AddIcon,
        DeleteIcon$1 = _useGlobalIcon.DeleteIcon,
        BrowseIcon$1 = _useGlobalIcon.BrowseIcon;

    var _useCommonClassName = useCommonClassName(),
        STATUS = _useCommonClassName.STATUS;

    var showTrigger = computed(function () {
      var multiple = props2.multiple,
          max = props2.max,
          files = props2.files;

      if (multiple) {
        return !max || files && files.length < max;
      }

      return !(files && files[0]);
    });

    var onMaskClick = function onMaskClick(e) {
      !showTrigger.value && props2.onClick(e);
    };

    var renderCardItem = function renderCardItem(file, index) {
      return createVNode("li", {
        "class": "".concat(UPLOAD_NAME.value, "__card-item ").concat(prefix.value, "-is--background")
      }, [createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__card-content ").concat(UPLOAD_NAME.value, "__card-box")
      }, [createVNode("img", {
        "class": "".concat(UPLOAD_NAME.value, "__card-image"),
        "src": file.url
      }, null), createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__card-mask"),
        "onClick": onMaskClick
      }, [createVNode("span", {
        "class": "".concat(UPLOAD_NAME.value, "__card-mask-item")
      }, [createVNode(BrowseIcon$1, {
        "onClick": function onClick(_ref) {
          var e = _ref.e;
          e.stopPropagation();
          props2.onImgPreview(e, file);
        }
      }, null)]), !disabled.value && [createVNode("span", {
        "class": "".concat(UPLOAD_NAME.value, "__card-mask-item-divider"),
        "key": "divider"
      }, null), createVNode("span", {
        "class": "".concat(UPLOAD_NAME.value, "__card-mask-item"),
        "key": "delete-icon"
      }, [createVNode(DeleteIcon$1, {
        "onClick": function onClick(_ref2) {
          var e = _ref2.e;
          e.stopPropagation();
          props2.onRemove({
            e: e,
            file: file,
            index: index
          });
        }
      }, null)])]])])]);
    };

    var renderTrigger = function renderTrigger() {
      var _props2$locale, _props2$locale$progre, _props2$locale2, _props2$locale2$trigg;

      return createVNode("li", {
        "class": ["".concat(UPLOAD_NAME.value, "__card-item ").concat(prefix.value, "-is--background"), _defineProperty({}, STATUS.value.disabled, disabled.value)],
        "onClick": props2.onClick
      }, [props2.showUploadProgress && props2.loadingFile && props2.loadingFile.status === "progress" ? createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__card-container ").concat(UPLOAD_NAME.value, "__card-box")
      }, [createVNode(Loading, null, null), createVNode("p", null, [((_props2$locale = props2.locale) === null || _props2$locale === void 0 ? void 0 : (_props2$locale$progre = _props2$locale.progress) === null || _props2$locale$progre === void 0 ? void 0 : _props2$locale$progre.uploadingText) || globalConfig.value.progress.uploadingText, " ", Math.min(props2.percent, 99), "%"])]) : createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__card-container ").concat(UPLOAD_NAME.value, "__card-box")
      }, [createVNode(AddIcon$1, null, null), createVNode("p", {
        "class": "".concat(prefix.value, "-size-s")
      }, [props2.placeholder || ((_props2$locale2 = props2.locale) === null || _props2$locale2 === void 0 ? void 0 : (_props2$locale2$trigg = _props2$locale2.triggerUploadText) === null || _props2$locale2$trigg === void 0 ? void 0 : _props2$locale2$trigg.image) || globalConfig.value.triggerUploadText.image])])]);
    };

    return function () {
      return createVNode("ul", {
        "class": "".concat(UPLOAD_NAME.value, "__card")
      }, [props2.files && props2.files.map(function (file, index) {
        return renderCardItem(file, index);
      }), showTrigger.value && renderTrigger()]);
    };
  }
});

export { ImageCard as default };
