// tdesign-vue-next/cjs/upload/dragger.js
/**
 * tdesign v0.21.1
 * (c) 2022 tdesign
 * @license MIT
 */

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _defineProperty = require('@babel/runtime/helpers/defineProperty');
var vue = require('vue');
var tdesignIconsVueNext = require('tdesign-icons-vue-next');
var button_index = require('../button/index.js');
var loading_index = require('../loading/index.js');
var upload_props = require('./props.js');
var upload_util = require('./util.js');
var hooks_tnode = require('../hooks/tnode.js');
var hooks_useConfig = require('../hooks/useConfig.js');
var hooks_useGlobalIcon = require('../hooks/useGlobalIcon.js');
var configProvider_useConfig = require('../config-provider/useConfig.js');
require('../button/button.js');
require('@babel/runtime/helpers/toConsumableArray');
require('../button/props.js');
require('../hooks/useRipple.js');
require('../hooks/useKeepAnimation.js');
require('../config-provider/context.js');
require('lodash/mergeWith');
require('lodash/merge');
require('../_common/js/global-config/default-config.js');
require('../_common/js/global-config/locale/zh_CN.js');
require('../utils/set-style.js');
require('../form/hooks.js');
require('../loading/directive.js');
require('../loading/plugin.js');
require('../loading/loading.js');
require('@babel/runtime/helpers/slicedToArray');
require('../loading/icon/gradient.js');
require('../_common/js/loading/circle-adapter.js');
require('../_common/js/utils/set-style.js');
require('../_common/js/utils/helper.js');
require('@babel/runtime/helpers/objectWithoutProperties');
require('../utils/dom.js');
require('../_chunks/dep-68610cdc.js');
require('lodash/isString');
require('../utils/easing.js');
require('../utils/render-tnode.js');
require('lodash/isEmpty');
require('lodash/isFunction');
require('lodash/isObject');
require('lodash/camelCase');
require('lodash/kebabCase');
require('../utils/transfer-dom.js');
require('../loading/props.js');
require('../utils/withInstall.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var _defineProperty__default = /*#__PURE__*/_interopDefaultLegacy(_defineProperty);

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty__default["default"](target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
var Dragger = vue.defineComponent({
  name: "TUploadDragger",
  props: {
    file: {
      type: Object,
      "default": function _default() {
        return null;
      }
    },
    loadingFile: {
      type: Object,
      "default": function _default() {
        return null;
      }
    },
    percent: {
      type: Number
    },
    locale: {
      type: Object,
      "default": function _default() {
        return null;
      }
    },
    autoUpload: upload_props["default"].autoUpload,
    theme: upload_props["default"].theme,
    onCancel: Function,
    onClick: Function,
    onRemove: Function,
    onUpload: Function,
    onChange: Function,
    onDragleave: Function,
    onDragenter: Function
  },
  setup: function setup(props2) {
    var renderTNodeJSX = hooks_tnode.useTNodeJSX();
    var target = vue.ref(null);
    var dragActive = vue.ref(false);

    var _useConfig = configProvider_useConfig.useConfig("upload"),
        globalConfig = _useConfig.globalConfig;

    var UPLOAD_NAME = hooks_useConfig.usePrefixClass("upload");

    var _useGlobalIcon = hooks_useGlobalIcon.useGlobalIcon({
      CheckCircleFilledIcon: tdesignIconsVueNext.CheckCircleFilledIcon,
      ErrorCircleFilledIcon: tdesignIconsVueNext.ErrorCircleFilledIcon
    }),
        CheckCircleFilledIcon = _useGlobalIcon.CheckCircleFilledIcon,
        ErrorCircleFilledIcon = _useGlobalIcon.ErrorCircleFilledIcon;

    var _useCommonClassName = hooks_useConfig.useCommonClassName(),
        SIZE = _useCommonClassName.SIZE;

    var imageUrl = vue.computed(function () {
      return props2.loadingFile && props2.loadingFile.url || props2.file && props2.file.url;
    });
    var inputName = vue.computed(function () {
      return props2.loadingFile && props2.loadingFile.name || props2.file && props2.file.name || "";
    });
    var classes = vue.computed(function () {
      return ["".concat(UPLOAD_NAME.value, "__dragger"), _defineProperty__default["default"]({}, "".concat(UPLOAD_NAME.value, "__dragger-center"), !props2.loadingFile && !props2.file), _defineProperty__default["default"]({}, "".concat(UPLOAD_NAME.value, "__dragger-error"), props2.loadingFile && props2.loadingFile.status === "fail")];
    });
    var size = vue.computed(function () {
      return props2.loadingFile && props2.loadingFile.size || props2.file && props2.file.size;
    });
    var showResultOperate = vue.computed(function () {
      var _props2$file, _props2$loadingFile;

      return Boolean(!props2.loadingFile && ((_props2$file = props2.file) === null || _props2$file === void 0 ? void 0 : _props2$file.name)) || ["success", "fail"].includes((_props2$loadingFile = props2.loadingFile) === null || _props2$loadingFile === void 0 ? void 0 : _props2$loadingFile.status);
    });

    var handleDrop = function handleDrop(event) {
      var _props2$onChange, _props2$onDragleave;

      event.preventDefault();
      (_props2$onChange = props2.onChange) === null || _props2$onChange === void 0 ? void 0 : _props2$onChange.call(props2, event.dataTransfer.files);
      (_props2$onDragleave = props2.onDragleave) === null || _props2$onDragleave === void 0 ? void 0 : _props2$onDragleave.call(props2, event);
      dragActive.value = false;
    };

    var handleDragenter = function handleDragenter(event) {
      var _props2$onDragenter;

      event.preventDefault();
      target.value = event.target;
      (_props2$onDragenter = props2.onDragenter) === null || _props2$onDragenter === void 0 ? void 0 : _props2$onDragenter.call(props2, event);
      dragActive.value = true;
    };

    var handleDragleave = function handleDragleave(event) {
      var _props2$onDragleave2;

      if (target.value !== event.target) return;
      event.preventDefault();
      (_props2$onDragleave2 = props2.onDragleave) === null || _props2$onDragleave2 === void 0 ? void 0 : _props2$onDragleave2.call(props2, event);
      dragActive.value = false;
    };

    var renderImage = function renderImage() {
      return vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-img-wrap")
      }, [imageUrl.value && vue.createVNode("img", {
        "src": imageUrl.value
      }, null)]);
    };

    var renderUploading = function renderUploading() {
      if (props2.loadingFile.status === "fail") return vue.createVNode(ErrorCircleFilledIcon, null, null);

      if (props2.loadingFile.status === "progress") {
        return vue.createVNode("div", {
          "class": "".concat(UPLOAD_NAME.value, "__single-progress")
        }, [vue.createVNode(loading_index.Loading, null, null), vue.createVNode("span", {
          "class": "".concat(UPLOAD_NAME.value, "__single-percent")
        }, [Math.min(props2.percent, 99), "%"])]);
      }
    };

    var renderProgress = function renderProgress() {
      var _props2$loadingFile2, _props2$locale, _props2$loadingFile3, _props2$locale2, _props2$locale2$trigg, _props2$locale3, _props2$locale3$trigg, _props2$locale4, _props2$locale4$trigg;

      return vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-progress")
      }, [props2.theme === "image" && renderImage(), vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-progress-info")
      }, [vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-text")
      }, [vue.createVNode("span", {
        "class": "".concat(UPLOAD_NAME.value, "__single-name")
      }, [upload_util.abridgeName(inputName.value)]), props2.loadingFile && renderUploading(), !props2.loadingFile && !!props2.file && vue.createVNode(CheckCircleFilledIcon, null, null)]), vue.createVNode("small", {
        "class": "".concat(SIZE.value.small)
      }, [globalConfig.value.file.fileSizeText, "\uFF1A", upload_util.returnFileSize(size.value)]), vue.createVNode("small", {
        "class": "".concat(SIZE.value.small)
      }, [globalConfig.value.file.fileOperationDateText, "\uFF1A", upload_util.getCurrentDate()]), vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-btns")
      }, [["progress", "waiting"].includes((_props2$loadingFile2 = props2.loadingFile) === null || _props2$loadingFile2 === void 0 ? void 0 : _props2$loadingFile2.status) && vue.createVNode(button_index.Button, {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(UPLOAD_NAME.value, "__dragger-progress-cancel"),
        "onClick": props2.onCancel
      }, {
        "default": function _default() {
          return [((_props2$locale = props2.locale) === null || _props2$locale === void 0 ? void 0 : _props2$locale.cancelUploadText) || globalConfig.value.cancelUploadText];
        }
      }), !props2.autoUpload && ((_props2$loadingFile3 = props2.loadingFile) === null || _props2$loadingFile3 === void 0 ? void 0 : _props2$loadingFile3.status) === "waiting" && vue.createVNode(button_index.Button, {
        "variant": "text",
        "theme": "primary",
        "onClick": function onClick(e) {
          return props2.onUpload(_objectSpread({}, props2.loadingFile), e);
        }
      }, {
        "default": function _default() {
          return [((_props2$locale2 = props2.locale) === null || _props2$locale2 === void 0 ? void 0 : (_props2$locale2$trigg = _props2$locale2.triggerUploadText) === null || _props2$locale2$trigg === void 0 ? void 0 : _props2$locale2$trigg.normal) || globalConfig.value.triggerUploadText.normal];
        }
      })]), showResultOperate.value && vue.createVNode("div", {
        "class": "".concat(UPLOAD_NAME.value, "__dragger-btns")
      }, [vue.createVNode(button_index.Button, {
        "theme": "primary",
        "variant": "text",
        "class": "".concat(UPLOAD_NAME.value, "__dragger-progress-cancel"),
        "onClick": function onClick(e) {
          props2.onRemove(e);
          props2.onClick(e);
        }
      }, {
        "default": function _default() {
          return [((_props2$locale3 = props2.locale) === null || _props2$locale3 === void 0 ? void 0 : (_props2$locale3$trigg = _props2$locale3.triggerUploadText) === null || _props2$locale3$trigg === void 0 ? void 0 : _props2$locale3$trigg.reupload) || globalConfig.value.triggerUploadText.reupload];
        }
      }), vue.createVNode(button_index.Button, {
        "theme": "danger",
        "variant": "text",
        "onClick": props2.onRemove
      }, {
        "default": function _default() {
          return [((_props2$locale4 = props2.locale) === null || _props2$locale4 === void 0 ? void 0 : (_props2$locale4$trigg = _props2$locale4.triggerUploadText) === null || _props2$locale4$trigg === void 0 ? void 0 : _props2$locale4$trigg["delete"]) || globalConfig.value.triggerUploadText["delete"]];
        }
      })])])]);
    };

    var renderDefaultDragElement = function renderDefaultDragElement() {
      var _props2$locale5, _props2$locale5$trigg;

      var unActiveElement = vue.createVNode("div", null, [vue.createVNode("span", {
        "class": "".concat(UPLOAD_NAME.value, "--highlight")
      }, [((_props2$locale5 = props2.locale) === null || _props2$locale5 === void 0 ? void 0 : (_props2$locale5$trigg = _props2$locale5.triggerUploadText) === null || _props2$locale5$trigg === void 0 ? void 0 : _props2$locale5$trigg.normal) || globalConfig.value.triggerUploadText.normal]), vue.createVNode("span", null, ["\xA0\xA0/\xA0\xA0", globalConfig.value.dragger.draggingText])]);

      var activeElement = vue.createVNode("div", null, [globalConfig.value.dragger.dragDropText]);

      return dragActive.value ? activeElement : unActiveElement;
    };

    return function () {
      var content = null;

      if ((props2.loadingFile || props2.file) && props2.theme !== "custom") {
        content = renderProgress();
      } else {
        content = vue.createVNode("div", {
          "class": "".concat(UPLOAD_NAME.value, "__trigger"),
          "onClick": props2.onClick
        }, [renderTNodeJSX("default") || renderDefaultDragElement()]);
      }

      return vue.createVNode("div", {
        "class": classes.value,
        "onDrop": handleDrop,
        "onDragenter": handleDragenter,
        "onDragleave": handleDragleave,
        "onDragover": function onDragover(event) {
          event.preventDefault();
        }
      }, [content]);
    };
  }
});

exports["default"] = Dragger;
